### PR TITLE
Implement more of the Query interface

### PR DESCRIPTION
**Pull request by darinf**
Adds `addSnapshotListener` and fleshes out `QuerySnapshot`.

### CODE REVIEW

File: Sources/FirebaseFirestore/NSError+FirestoreError.swift [hunk 1]
```diff
@@ -5,9 +5,14 @@ extension NSError {
...
   /// - Parameter error: The error returned from Firestore
   /// - Returns: An `NSError` if the underlying error code is non-zero.
-  internal static func firestore(_ error: firebase.firestore.Error?) -> NSError? {
+  internal static func firestore(_ error: firebase.firestore.Error?, errorMessage: UnsafePointer<CChar>? = nil) -> NSError? {
     guard let actualError = error, actualError.rawValue != 0 else { return nil }
 
-    return NSError(domain: "firebase.firestore", code: Int(actualError.rawValue))
+    var userInfo = [String: Any]()
+    if let errorMessage {
+      userInfo[NSLocalizedDescriptionKey] = String(cString: errorMessage)
```
Comment (brianmichel) [inline Sources/FirebaseFirestore/NSError+FirestoreError.swift:13]: nice addition!

File: Sources/FirebaseFirestore/Query+Swift.swift [hunk 2]
```diff
@@ -39,4 +39,147 @@ extension Query {
...
+
+      // Make sure we dispatch our callback back into the main thread to keep consistent
+      // with the reference API which will call back on the 'user_executor' which typically
+      // ends up being the main queue.
+      // Relevant code:
+      // - https://github.com/firebase/firebase-ios-sdk/blob/main/Firestore/Source/API/FIRFirestore.mm#L210-L218
+      // - https://github.com/firebase/firebase-ios-sdk/blob/main/Firestore/core/src/api/document_reference.cc#L236-L237
+      DispatchQueue.main.async {
+        callback(returned, error)
+      }
```
Comment (brianmichel) [inline Sources/FirebaseFirestore/Query+Swift.swift]: This was specifically for `DocumentReference` do we know if the same assumptions about which return queue is used can hold true for Query?

File: Sources/FirebaseFirestore/QuerySnapshot+Swift.swift [hunk 3]
```diff
@@ -6,3 +6,37 @@ import firebase
...
+
+  public var documentChanges: [DocumentChange] {
+    Array<DocumentChange>(
+      swift_firebase.swift_cxx_shims.firebase.firestore.query_snapshot_document_changes(self, .exclude)
+    )
+  }
+
+  public func documentChanges(includeMetadataChanges: Bool) -> [DocumentChange] {
+    Array<DocumentChange>(
+      swift_firebase.swift_cxx_shims.firebase.firestore.query_snapshot_document_changes(self, includeMetadataChanges ? .include : .exclude)
```
Comment (brianmichel) [inline Sources/FirebaseFirestore/QuerySnapshot+Swift.swift:39]: It looks like `query_snapshot_document_changes` includes a default value of `.exclude` in the C++ definition within `FirebaseFirestore.hh` should we reflect that default in this function definition as well? Probably not totally needed, but wanted to ask.

**Review by brianmichel — APPROVED** [review]
looks great, only a few small comments/questions

File: Sources/FirebaseFirestore/Query+Swift.swift (hunk 2 above)
Comment (darinf) [reply Sources/FirebaseFirestore/Query+Swift.swift]: I read the code and tested it. The callbacks in the Obj C API are intended to run on the main queue, but can be overridden thru FirebaseSettings.

File: Sources/FirebaseFirestore/QuerySnapshot+Swift.swift (hunk 3 above)
Comment (darinf) [reply Sources/FirebaseFirestore/QuerySnapshot+Swift.swift:39]: No, the strategy I took is to declare a computed property `documentChanges` to align with the way the Obj C API is reflected to Swift. Like the Obj C API, this override is a function instead for the extra parameter.

**Pull request merged**